File: app/src/main/java/xyz/zpayh/myadapter/MainActivity.java
package xyz.zpayh.myadapter;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.View;

import java.util.ArrayList;
import java.util.List;

import xyz.zpayh.adapter.BaseAdapter;
import xyz.zpayh.adapter.BaseViewHolder;
import xyz.zpayh.adapter.OnItemClickListener;

public class MainActivity extends AppCompatActivity {

    private RecyclerView mRecyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initView();
    }

    private void initView() {
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        mRecyclerView = (RecyclerView) findViewById(R.id.list);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));

        final BaseAdapter<String> adapter = new BaseAdapter<String>() {
            @Override
            public int getLayoutRes(int index) {
                return R.layout.item_list;
            }

            @Override
            public void convert(BaseViewHolder holder, String data, int index) {
                holder.setText(R.id.tv_act_title,data);
            }

            @Override
            public void bind(BaseViewHolder holder, int layoutRes) {
                holder.setClickable(R.id.app_root,true);
            }
        };
        adapter.setOnItemClickListener(new OnItemClickListener() {
            @Override
            public void onItemClick(@NonNull View view, int adapterPosition) {
                String action = adapter.getData(adapterPosition);
                Intent intent = new Intent();
                intent.setAction(action);
                startActivity(intent);
            }
        });

        mRecyclerView.setAdapter(adapter);

        String[] list = getResources().getStringArray(R.array.activity_title);
        final List<String> data = new ArrayList<>(list.length);
        for (String s : list) {
            data.add(s);
        }
        adapter.setData(data);
    }
}

File: sample-kotlin/src/main/java/cn/sherlockzp/sample/MainActivity.kt
package cn.sherlockzp.sample

import android.content.ComponentName
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import android.view.View
import cn.sherlockzp.adapter.BaseAdapter
import cn.sherlockzp.adapter.BaseViewHolder
import cn.sherlockzp.adapter.OnItemClickListener
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private val adapter = object : BaseAdapter<ResolveInfo>() {
        override fun getLayoutRes(index: Int) = R.layout.item_activity

        override fun convert(holder: BaseViewHolder, data: ResolveInfo, index: Int) {
            holder.setText(R.id.tv_activity_title, data.activityInfo.labelRes)
                    //.setText(R.id.tv_activity_label,"测试啦%s,%d","NB",index)
                    .setText(R.id.tv_activity_label,R.string.format_label,"NB",index)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        rv_list_activity.layoutManager = LinearLayoutManager(this)
        rv_list_activity.adapter = adapter

        adapter.onItemClickListener = object : OnItemClickListener{
            override fun onItemClick(view: View, adapterPosition: Int) {
                val resolveInfo = adapter.getData(adapterPosition)
                resolveInfo?.let {
                    val intent = Intent()
                    intent.action = "Sherlock"
                    intent.component = ComponentName(it.activityInfo.packageName,it.activityInfo.name)
                    startActivity(intent)
                }
            }
        }

        initData()

        rv_list_activity.postDelayed({
            adapter.doNotifyItemChanged(1)
        },1000)
    }

    private fun initData() {
        val intent = Intent()
        intent.action = "Sherlock"
        val list = packageManager.queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY)
        adapter.setData(list)
    }
}
